#2424 mixed usage of rest url

BE-Webdesign opened the pull request:
Should fix #2424, however, after writing a unit test it doesn't appear to matter whether the '/' is there or not as get_rest_url adds a leading slash regardless as far as I can tell.  Also it is very difficult to write this kind of test so any help would be appreciated.

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into develop

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into develop

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into develop

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into develop

BE-Webdesign committed:
#2424-mixed-usage-of-rest-url

Should fix #2424, however, after writing a unit test it doesn't appear
to matter whether the '/' is there or not as get_rest_url adds a leading
slash regardless as far as I can tell.

danielbachhuber commented on tests/class-wp-test-rest-controller-testcase.php:
```diff
@@ -6,7 +6,7 @@
 	public function setUp() {
 		parent::setUp();
-
+		add_filter( 'rest_url', array( $this, 'test_rest_url_for_leading_slash' ), 10, 2 );
```
I think you need to remove this filter on `tearDown` too

danielbachhuber commented on tests/class-wp-test-rest-controller-testcase.php:
```diff
@@ -39,4 +39,8 @@ public function tearDown() {
 
+	public function test_rest_url_for_leading_slash( $url, $path ) {
+		// Make sure path for rest_url has a leading slash for proper resolution.
+		$this->assertTrue( 0 === strpos( $path, '/' ) );
```
Because you're running this test on a filter, you need to make sure to return the `$url` :) Otherwise, the sky will fall upon your head, as it has done.

BE-Webdesign replied:
Yeah, I thought of that not sure why I never put it in.  The last thing is that this function is called on every subclass as though it is a test so you get an exception thrown because no parameters are passed into it.  I tried to make it protected but then the subclasses can't access it.  Should I change this to parent in the filter?

danielbachhuber replied:
You can rename it to `filter_rest_url_for_leading_slash`

BE-Webdesign replied:
Thanks, that solved everything.  One last thing though, the test itself doesn't really help at all for what we are trying to do as `get_rest_url()` always adds the leading slash before it hits the filter `rest_url`.  How would we test the input into the filter to see if there is a slash or not?  Currently the test itself just checks to make sure there is a leading slash at the start of $path and asserts whether that is true which it is regardless of whether there is a leading slash or not being put into the `rest_url()`.

danielbachhuber replied:
> One last thing though, the test itself doesn't really help at all for what we are trying to do as get_rest_url() always adds the leading slash before it hits the filter rest_url.

Oh. @timmyc How are you running into your problem then?

BE-Webdesign replied:
I think they are rewriting `rest_url()`, from my understanding.

BE-Webdesign committed:
Fixing tests.

Thank you danielbachuber.  Test now works, but I am not sure if the test
should work differently.  Currently it only checks if a leading slash is
on the path which it should be.  However, it is difficult to get the
untouched data checked to see if it has a leading slash or not, which is
the intention of the test.

joehoyle commented:
This looks good to me, I think we can roll this in as-is. Thanks @BE-Webdesign!
